File: app/src/main/java/se/ntlv/newsbringer/view/DateView.java
package se.ntlv.newsbringer.view;

import android.content.Context;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.widget.TextView;

public class DateView extends TextView {

	public DateView(Context context, AttributeSet attributeSet) {
		super(context, attributeSet);
	}

	static CharSequence reformat(CharSequence text) {
		try {
			long asLong = Long.parseLong(String.valueOf(text));
			long diff = System.currentTimeMillis() / 1000 - asLong;
			long dividend = (3600 > diff) ? 60 : 3600;
			String suffix = (3600 > diff) ? "min" : "hours";
			return (diff / dividend) + suffix;

		} catch (NumberFormatException e) {
			return text;
		}
	}

	public void setText(CharSequence text, TextView.BufferType type) {
		if (TextUtils.isEmpty(text)) {
			return;
		}
		super.setText(reformat(text), type);
	}
}

File: app/src/main/kotlin/se/ntlv/newsbringer/DateView.kt
package se.ntlv.newsbringer

import android.content.Context
import android.util.AttributeSet
import android.widget.TextView

class DateView: TextView {
    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle)
    constructor(context: Context, attrs: AttributeSet) : super(context, attrs, 0)
    constructor(context: Context) : super(context)

    override fun setText(text: CharSequence?, type: TextView.BufferType?) {
        if ( isInEditMode) {
            super.setText(text, type)
            return
        }
        val reformattedText = text?.reformat()
        super.setText(reformattedText ?: text, type)
    }
    fun CharSequence.reformat(): CharSequence {
        try {
            val asLong = this.toString().toLong()
            val diff = System.currentTimeMillis().div(1000).minus(asLong)
            return when {
                diff < 60L -> formatTime(diff, TimeUnit.SECOND)
                diff < 3600L -> formatTime(diff, TimeUnit.MINUTE)
                diff < 86400L -> formatTime(diff, TimeUnit.HOUR)
                else -> formatTime(diff, TimeUnit.DAY)
            }
        } catch (formatException: NumberFormatException) {
            return this;
        }
    }

    enum class TimeUnit(val dividend: Long, val singular: Int, val plural: Int) {
        SECOND(1L, R.string.second, R.string.seconds),
        MINUTE(60L, R.string.minute, R.string.minutes),
        HOUR(3600L, R.string.hour, R.string.hours),
        DAY(86400L, R.string.day, R.string.days)
    }

    fun getString(id : Int) : String = context.getString(id)

    fun formatTime(diff: Long, unit: TimeUnit): CharSequence {
        val time = diff.div(unit.dividend)
        return when (time) {
            1L -> "1 ${getString(unit.singular)}"
            else -> "$time ${getString(unit.plural)}"
        }
    }

}
